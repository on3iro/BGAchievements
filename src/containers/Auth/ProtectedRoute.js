import React from 'react'
import PropTypes from 'prop-types'
import { Route, Redirect } from 'react-router-dom'
import { connect } from 'react-redux'

import { getLoggedIn } from './ducks/selectors'

export const ProtectedRoute = ({ component, ...rest }) => (
  <Route {...rest} render={props => (
    rest.isLoggedIn
      ? (
      React.createElement(component, props)
    ) : (
      <Redirect to={{
        pathname: '/login',
        state: { from: props.location }
      }} />
    )
  )} />
)

export const mapStateToProps = (state) => {
  return {
    isLoggedIn: getLoggedIn(state)
  }
}

ProtectedRoute.propTypes = {
  component: PropTypes.any.isRequired,
  location: PropTypes.object
}

export default connect(mapStateToProps, null)(ProtectedRoute)
