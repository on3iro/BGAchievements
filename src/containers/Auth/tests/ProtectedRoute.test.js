import React from 'react'
import { mount } from 'enzyme'
import { Redirect, MemoryRouter } from 'react-router-dom'

import { ProtectedRoute, mapStateToProps } from '../ProtectedRoute'

const renderComponent = newProps => {
  const TestComponent = () => {
    return (
      <div>test</div>
    )
  }
  const props = {
    component: TestComponent,
    isLoggedIn: true,
    ...newProps
  }

  const renderedWrapper = mount(
    <MemoryRouter>
      <ProtectedRoute {...props} />
    </MemoryRouter>
  )

  return {
    props,
    renderedWrapper,
    TestComponent
  }
}

describe('<ProtectedRoute />', () => {
  it('should render specified component', () => {
    const { renderedWrapper, TestComponent } = renderComponent()
    expect(renderedWrapper.find(TestComponent).length).toMatchSnapshot()
  })

  it('should render Redirect to "/login" if user is logged out', () => {
    const { renderedWrapper } = renderComponent({ isLoggedIn: false })
    const redirect = renderedWrapper.find(Redirect)

    expect(redirect.length).toMatchSnapshot()
    expect(redirect.props().to.pathname).toMatchSnapshot()
  })

  describe('react-redux', () => {
    it('should map state to props', () => {
      expect(mapStateToProps({
        auth: { user: { isLoggedIn: true }}
      })).toMatchSnapshot()
    })
  })
})
