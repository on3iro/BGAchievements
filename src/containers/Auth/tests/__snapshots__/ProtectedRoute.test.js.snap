exports[`<ProtectedRoute /> react-redux should map state to props 1`] = `
Object {
  "isLoggedIn": true,
}
`

exports[`<ProtectedRoute /> should render Redirect to "/login" if user is logged out 1`] = `1`

exports[`<ProtectedRoute /> should render Redirect to "/login" if user is logged out 2`] = `"/login"`

exports[`<ProtectedRoute /> should render specified component 1`] = `1`
